### PR TITLE
Adding FLNActivationQuantizationHolder and PreservingActivationQuantizationHolder

### PR DESCRIPTION
**Pull request by kkawa14**
## Pull Request Description:
Adding a PytorchFLNActivationQuantizationHolder class and including tests.
Adding a PytorchPreservingActivationQuantizationHolder class and including tests.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: tests/pytorch_tests/test_preserving_activation_quantizer_holder.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+import torch
+import numpy as np
+from torch.fx import symbolic_trace
+
+from mct_quantizers.pytorch.preserving_activation_quantization_holder import PytorchPreservingActivationQuantizationHolder
+from mct_quantizers.pytorch.quantizers import ActivationSymmetricInferableQuantizer, ActivationPOTInferableQuantizer, ActivationUniformInferableQuantizer
+
+
+class TestPytorchPreservingActivationQuantizationHolderInference(unittest.TestCase):
```
Comment (elad-c) [inline tests/pytorch_tests/test_preserving_activation_quantizer_holder.py:27]: In general:
* Add bypass tests.
* Same comments as for `tests/pytorch_tests/test_fln_activation_quantizer_holder.py`

**Review by elad-c — CHANGES_REQUESTED** [review]

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+            Args:
+                activation_holder_quantizer: Quantizer to use during inference.
+                quantization_bypass: Indicates whether to bypass quantization for the activation holder.
+                **kwargs: Key-word arguments for the base layer
+            """
+
+            super(PytorchFLNActivationQuantizationHolder, self).__init__(activation_holder_quantizer=activation_holder_quantizer,
+                                                                         **kwargs)
+            self.activation_holder_quantizer.initialize_quantization(None,
```
Comment (elad-c) [inline mct_quantizers/pytorch/fln_activation_quantization_holder.py:42]: Can remove this since `initialize_quantization` actually does nothing.

File: mct_quantizers/pytorch/preserving_activation_quantization_holder.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            """
+
+            Args:
+                activation_holder_quantizer: Quantizer to use during inference.
+                **kwargs: Key-word arguments for the base layer
+            """
+
+            super(PytorchPreservingActivationQuantizationHolder, self).__init__(activation_holder_quantizer=activation_holder_quantizer,
+                                                                                **kwargs)
+            self.activation_holder_quantizer.initialize_quantization(None,
```
Comment (elad-c) [inline mct_quantizers/pytorch/preserving_activation_quantization_holder.py:40]: Can remove this since `initialize_quantization` actually does nothing.

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Fixed it for '2025'.

File: mct_quantizers/pytorch/preserving_activation_quantization_holder.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply mct_quantizers/pytorch/preserving_activation_quantization_holder.py]: Fixed it for '2025'.

File: mct_quantizers/pytorch/preserving_activation_quantization_holder.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+from mct_quantizers.common.base_inferable_quantizer import BaseInferableQuantizer
+from mct_quantizers.common.constants import FOUND_TORCH, PRESERVING_ACTIVATION_HOLDER_QUANTIZER
+from mct_quantizers.pytorch.activation_quantization_holder import PytorchActivationQuantizationHolder
+from mct_quantizers.logger import Logger
+
+if FOUND_TORCH:
+    import torch
+
+    class PytorchPreservingActivationQuantizationHolder(PytorchActivationQuantizationHolder):
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply mct_quantizers/pytorch/preserving_activation_quantization_holder.py:24]: Adding a bypass flag to this class.

File: tests/pytorch_tests/test_fln_activation_quantizer_holder.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply tests/pytorch_tests/test_fln_activation_quantizer_holder.py]: Fixed it for '2025'.

File: tests/pytorch_tests/test_preserving_activation_quantizer_holder.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+# Copyright 2023 Sony Semiconductor Israel, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply tests/pytorch_tests/test_preserving_activation_quantizer_holder.py]: Fixed it for '2025'.

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        def __init__(self,
+                     activation_holder_quantizer: BaseInferableQuantizer,
+                     quantization_bypass: bool = False,
+                     **kwargs):
+            """
+
+            Args:
+                activation_holder_quantizer: Quantizer to use during inference.
+                quantization_bypass: Indicates whether to bypass quantization for the activation holder.
+                **kwargs: Key-word arguments for the base layer
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Corrected the argument description of kwargs.
Should I also fix the message for PytorchActivationQuantizationHolder in this pull request?

File: mct_quantizers/pytorch/preserving_activation_quantization_holder.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        Pytorch module to hold an activation quantizer and quantize during inference.
+        """
+        def __init__(self,
+                     activation_holder_quantizer: BaseInferableQuantizer,
+                     **kwargs):
+            """
+
+            Args:
+                activation_holder_quantizer: Quantizer to use during inference.
+                **kwargs: Key-word arguments for the base layer
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply mct_quantizers/pytorch/preserving_activation_quantization_holder.py]: Corrected the argument description of kwargs.

File: tests/pytorch_tests/test_preserving_activation_quantizer_holder.py (hunk 1 above)
Comment (kkawa14) [reply tests/pytorch_tests/test_preserving_activation_quantizer_holder.py:28]: Made the same fix as `test_fln_activation_quantizer_holder.py`

File: tests/pytorch_tests/test_fln_activation_quantizer_holder.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                
+                with tempfile.NamedTemporaryFile(delete=False, suffix='.pth') as tmp_file:
+                    tmp_pth_file = tmp_file.name
+                
+                try:
+                    torch.save(fx_model, tmp_pth_file)
+                    loaded_model = torch.load(tmp_pth_file)
+                    loaded_model(x)
+                finally:
+                    os.remove(tmp_pth_file)
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply tests/pytorch_tests/test_fln_activation_quantizer_holder.py:128]: Added a test for the output of quantization holder before and after saving.

File: tests/pytorch_tests/test_fln_activation_quantizer_holder.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                self.assertTrue(np.allclose(model.activation_holder_quantizer.max_range, exp_quantizer_args["max_range"][0]))
+                self.assertTrue(np.allclose(model.activation_holder_quantizer.scale, exp_quantizer_args["scale"][0]))
+            else:
+                self.assertTrue(model.activation_holder_quantizer.num_bits == exp_quantizer_args["num_bits"])
+                self.assertTrue(model.activation_holder_quantizer.threshold_np == exp_quantizer_args["threshold"])
+                self.assertTrue(model.activation_holder_quantizer.signed == exp_quantizer_args["signed"])
+
+                if not quantization_bypass:
+                    # The maximal threshold is 4 using a signed quantization, so we expect all values to be between -4 and 4
+                    self.assertTrue(quantized_tensor.max().item() < exp_quantizer_args["threshold"][0], f'Quantized values should not contain values greater than maximal threshold ')
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply tests/pytorch_tests/test_fln_activation_quantizer_holder.py:78]: Added a test for comparing the output between `PytorchFLNActivationQuantizationHolder` and `PytorchActivationQuantizationHolder` in L84-L86.

File: tests/pytorch_tests/test_fln_activation_quantizer_holder.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+                    self.assertTrue(quantized_tensor.min().item() >= -exp_quantizer_args["threshold"][0], f'Quantized values should not contain values lower than minimal threshold ')
+
+            if quantization_bypass:
+                # Value is the same for input and output
+                self.assertTrue(np.allclose(quantized_tensor, input_tensor), f'Expected values are the same tensor but output tensor is {quantized_tensor}')  
+            else:
+                self.assertTrue(len(torch.unique(quantized_tensor)) <= 2 ** quantizer_args["num_bits"], f'Quantized tensor expected to have no more than {2 ** quantizer_args["num_bits"]} unique values but has {len(np.unique(quantized_tensor))} unique values')
```
[the remark this answers is not in the document]
Comment (kkawa14) [reply tests/pytorch_tests/test_fln_activation_quantizer_holder.py:88]: Added a test for comparing the output between `PytorchFLNActivationQuantizationHolder` and `PytorchActivationQuantizationHolder` in L84-L86.

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py (hunk 2 above)
Comment (kkawa14) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Removed the relevant section.
Should the additional Constant variables used in the removed section also be deleted from `common/constant.py`?

File: mct_quantizers/pytorch/preserving_activation_quantization_holder.py (hunk 3 above)
Comment (kkawa14) [reply mct_quantizers/pytorch/preserving_activation_quantization_holder.py]: Removed the relevant section.

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (elad-c) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Yes, please

**Review by elad-c — APPROVED** [review]

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py (hunk 2 above)
Comment (elad-c) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Yes please

Comment (kkawa14) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Removed unused constant from `common/constant.py`.

File: mct_quantizers/pytorch/fln_activation_quantization_holder.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kkawa14) [reply mct_quantizers/pytorch/fln_activation_quantization_holder.py]: Corrected the argument description of kwargs for `PytorchActivationQuantizationHolder`.
And added a test for the output of quantization holder before and after saving.

**Comment by ofirgo** [comment]
Hi @kkawa14 ,

Note that the PR validation tests are failing.
This is due to a packaging issue that is unrelated to your changes.
This is fixed on main, so, please sync main into your branch so we can run the tests again and verify that everything works correctly and we can merge your PR.
